src/category_theory/abelian/pseudoelements.lean

```diff
@@ -0,0 +1,411 @@
+
+section
+local attribute [instance] has_pullbacks
+
+/-- An epimorphism is surjective on pseudoelements. -/
+theorem pseudo_surjective_of_epi {P Q : C} (f : P ⟶ Q) [epi f] : function.surjective f :=
+λ qbar, quotient.induction_on qbar $ λ q, ⟨((pullback.fst : pullback f q.hom ⟶ P) : over P),
+  quotient.sound $ ⟨pullback f q.hom, 𝟙 (pullback f q.hom), pullback.snd, by apply_instance,
+    by apply_instance, by rw [category.id_comp, ←pullback.condition, app_hom, over.coe_hom]⟩⟩
+
+end
+
+/-- A morphism that is surjective on pseudoelements is an epimorphism. -/
+theorem epi_of_pseudo_surjective {P Q : C} (f : P ⟶ Q) : function.surjective f → epi f :=
+λ h, match h (𝟙 Q) with ⟨pbar, hpbar⟩ :=
+  match quotient.exists_rep pbar with ⟨p, hp⟩ :=
+    have ⟦(p.hom ≫ f : over Q)⟧ = ⟦𝟙 Q⟧, by { rw ←hp at hpbar, exact hpbar },
+    match quotient.exact this with ⟨R, x, y, ex, ey, comm⟩ :=
+      @epi_of_epi_fac _ _ _ _ _ (x ≫ p.hom) f y ey $
+        by { dsimp at comm, rw [category.assoc, comm], apply category.comp_id }
+    end
+  end
+end
+
+section
+local attribute [instance] preadditive.has_equalizers_of_has_kernels
+local attribute [instance] has_pullbacks
+
+/-- Two morphisms in an exact sequence are exact on pseudoelements. -/
+theorem pseudo_exact_of_exact {P Q R : C} {f : P ⟶ Q} {g : Q ⟶ R} [exact f g] :
+  (∀ a, g (f a) = 0) ∧ (∀ b, g b = 0 → ∃ a, f a = b) :=
+⟨λ a, by { rw [←comp_apply, exact.w], exact zero_apply _ _ },
+  λ b', quotient.induction_on b' $ λ b hb,
+    have hb' : b.hom ≫ g = 0, from (pseudo_zero_iff _).1 hb,
+    begin
+      -- By exactness, b factors through im f = ker g via some c
+      obtain ⟨c, hc⟩ := kernel_fork.is_limit.lift' (is_limit_image f g) _ hb',
+
+      -- We compute the pullback of the map into the image and c.
+      -- The pseudoelement induced by the first pullback map will be our preimage.
+      use (pullback.fst : pullback (factor_thru_image f) c ⟶ P),
+
+      -- It remains to show that the image of this element under f is pseudo-equal to b.
+      apply quotient.sound,
+
+      -- pullback.snd is an epimorphism because the map onto the image is!
+      refine ⟨pullback (factor_thru_image f) c, 𝟙 _, pullback.snd,
+        by apply_instance, by apply_instance, _⟩,
+
+      -- Now we can verify that the diagram commutes.
+      calc 𝟙 (pullback (factor_thru_image f) c) ≫ pullback.fst ≫ f = pullback.fst ≫ f
+                : category.id_comp _
+        ... = pullback.fst ≫ factor_thru_image f ≫ kernel.ι (cokernel.π f)
+                : by rw [kernel_cokernel_eq_image_ι, ←image_ι_eq_image_ι, image.fac]
+        ... = (pullback.snd ≫ c) ≫ kernel.ι (cokernel.π f)
+                : by rw [←category.assoc, pullback.condition]
+        ... = pullback.snd ≫ b.hom
+                : by { rw category.assoc, congr,
+                       simpa [image_ι_eq_image_ι, kernel_cokernel_eq_image_ι] using hc }
+    end⟩
+
+end
+
+lemma comp_zero {P Q R : C} (f : Q ⟶ R) (a : P ⟶ Q) : a ≫ f = 0 → f a = 0 :=
```

Comment:
```suggestion
lemma apply_eq_zero_of_comp_eq_zero {P Q R : C} (f : Q ⟶ R) (a : P ⟶ Q) : a ≫ f = 0 → f a = 0 :=
```
I dunno... it's a trade off between brevity and descriptiveness... Maybe yours is better. Just a suggestion.